lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,855 @@
+        self.current_message_length += chunk.remaining();
+        self.chunks[sequence_number as usize] = chunk;
+    }
+
+    fn is_complete(&self) -> bool {
+        self.chunks_bitmap.count_ones() == self.total_chunks as u32
+    }
+
+    fn current_message_length(&self) -> usize {
+        self.current_message_length
+    }
+
+    fn retrieve_message(&self) -> Option<Bytes> {
+        if self.is_complete() {
+            self.timeout_task.abort();
+            let chunks = &self.chunks[0..self.total_chunks as usize];
+            let mut message = BytesMut::new();
+            for chunk in chunks {
+                message.extend_from_slice(chunk);
+            }
+            Some(message.freeze())
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Debug, Snafu, PartialEq, Eq)]
+pub enum ChunkedGelfDecoderError {
+    #[snafu(display("Invalid chunk header with less than 10 bytes: 0x{header:0x}"))]
+    InvalidChunkHeader { header: Bytes },
+    #[snafu(display("Received chunk with message id {message_id} and sequence number {sequence_number} has an invalid total chunks value of {total_chunks}. It must be between 1 and {GELF_MAX_TOTAL_CHUNKS}."))]
+    InvalidTotalChunks {
+        message_id: u64,
+        sequence_number: u8,
+        total_chunks: u8,
+    },
+    #[snafu(display("Received chunk with message id {message_id} and sequence number {sequence_number} has a sequence number greater than its total chunks value of {total_chunks}"))]
+    InvalidSequenceNumber {
+        message_id: u64,
+        sequence_number: u8,
+        total_chunks: u8,
+    },
+    #[snafu(display("Pending messages limit of {pending_messages_limit} reached while processing chunk with message id {message_id} and sequence number {sequence_number}"))]
+    PendingMessagesLimitReached {
+        message_id: u64,
+        sequence_number: u8,
+        pending_messages_limit: usize,
+    },
+    #[snafu(display("Received chunk with message id {message_id} and sequence number {sequence_number} has different total chunks values: original total chunks value is {original_total_chunks} and received total chunks value is {received_total_chunks}"))]
+    TotalChunksMismatch {
+        message_id: u64,
+        sequence_number: u8,
+        original_total_chunks: u8,
+        received_total_chunks: u8,
+    },
+    #[snafu(display("Received chunk with message id {message_id} and sequence number {sequence_number} has exceeded the maximum chunk length: got {chunk_length} bytes and max chunk length is {max_chunk_length} bytes"))]
+    MaxChunkLengthExceeded {
+        message_id: u64,
+        sequence_number: u8,
+        chunk_length: usize,
+        max_chunk_length: usize,
+    },
+    #[snafu(display("Message with id {message_id} has exceeded the maximum message length and it will be dropped: got {message_length} bytes and max message length is {max_message_length} bytes. Discarding all buffered chunks of that message"))]
```

Comment:
@jszwedko Note that the discarding notice to the user is done inside this error. It will be logged outside of this framer when the `MaxMessageLengthExceede` error is thrown. This is a bit different from what you depicted about https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/lib/codecs/src/decoding/framing/character_delimited.rs#L123-L128 where the message is silently discarded and no error is returned to the caller.
What do you prefer? To add a discarding warning or an error? In my opinion, this is not really an error and fits more a warning, but as data is discarded, you said in previous comments that we should return an error in those cases.

Reply:
That's a good question. This has me wondering what the effect of returning an error from `decode()` is. Do you know if it kills the connection? Or if it just emits the error and continues decoding? If it is the latter, than I think returning an error here is appropriate (and we should probably update the other framers to do the same). If it kills the connection, we may want to avoid returning an error unless it is unrecoverable and simply emit warnings 🤔

Reply:
I don't think the connection is killed. Here https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L175
there is an infinite loop where at each iteration, `socket.recv_from(..)` is called to read an UDP datagram. If any of the `FramedRead` calls returns an error https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L202 
It will be handled by this case
https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L275

But that `FramedRead` stream won't return more than 1 result, as a chunked GELF datagram can't resolve to multiple GELF messages in a single call. The `!error.can_continue()` call is useless in this case, as every datagram would resolve to either: 1 single GELF message (if the chunks that is being processed is the last of the mesasge, no results at all, or an error. In every case, a new iteration of the infinite loop starts again, to read another datagram from the socket at https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L175C24-L175C40

I hope I'm explaining it well.

TLDR; Does not kill the connection, it continues decoding next datagrams on the socket.